internal/keystore/entrust/keycontrol.go

```diff
@@ -0,0 +1,575 @@
+		ids, prefix, err = kc.list(ctx, prefix, 250)
+		if err != nil {
+			return nil, err
+		}
+		names = append(names, ids...)
+
+		if prefix == "" || len(ids) == 0 {
+			break
+		}
+	}
+	return &iter{names: names}, nil
+}
+
+func (kc *KeyControl) list(ctx context.Context, prefix string, n int) ([]string, string, error) {
+	const (
+		Method     = http.MethodPost
+		Path       = "/vault/1.0/ListSecretIds/"
+		VaultToken = "X-Vault-Auth"
+	)
+	type Token struct {
+		BoxID      string `json:"box_id"`
+		ContinueAt string `json:"next_ctx"`
+	}
+	type Request struct {
+		BoxID     string `json:"box_id"`
+		N         int    `json:"max_items"`
+		NextToken string `json:"next_token,omitempty"` // Must be omitted if empty
+	}
+	type Response struct {
+		Secrets []struct {
+			Name    string `json:"name"`
+			Expired bool   `json:"expired"`
+		}
+		NextToken string `json:"next_token"`
+	}
+
+	var token string
+	if prefix != "" {
+		b, err := json.Marshal(Token{
+			BoxID:      kc.config.BoxID,
+			ContinueAt: prefix,
+		})
+		if err != nil {
+			return nil, "", fmt.Errorf("keycontrol: failed to list keys: %v", err)
+		}
+		token = base64.StdEncoding.EncodeToString(b)
+	}
+	body, err := json.Marshal(Request{
+		BoxID:     kc.config.BoxID,
+		N:         n,
+		NextToken: token,
+	})
+	if err != nil {
+		return nil, "", fmt.Errorf("keycontrol: failed to list keys: %v", err)
+	}
+
+	url, err := url.JoinPath(kc.config.Endpoint, Path)
+	if err != nil {
+		return nil, "", fmt.Errorf("keycontrol: failed to list keys: %v", err)
+	}
+	req, err := http.NewRequestWithContext(ctx, Method, url, xhttp.RetryReader(bytes.NewReader(body)))
+	if err != nil {
+		return nil, "", fmt.Errorf("keycontrol: failed to list keys: %v", err)
+	}
+	req.ContentLength = int64(len(body))
+	req.Header.Set(VaultToken, *kc.token.Load())
+
+	resp, err := kc.client.Do(req)
+	if err != nil {
+		return nil, "", fmt.Errorf("keycontrol: failed to list keys: %v", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, "", parseErrorResponse(resp)
+	}
+
+	var response Response
+	if err := json.NewDecoder(mem.LimitReader(resp.Body, 10*mem.MB)).Decode(&response); err != nil {
+		return nil, "", fmt.Errorf("keycontrol: failed to list keys: %v", err)
+	}
+	names := make([]string, len(response.Secrets))
+	for _, secret := range response.Secrets {
+		if !secret.Expired {
+			names = append(names, secret.Name)
+		}
+	}
+	if response.NextToken != "" {
+		rawToken, err := base64.StdEncoding.DecodeString(response.NextToken)
+		if err != nil {
+			return nil, "", fmt.Errorf("keycontrol: failed to list keys: invalid continue token: %v", err)
+		}
+
+		var token Token
+		if err = json.Unmarshal(rawToken, &token); err != nil {
+			return nil, "", fmt.Errorf("keycontrol: failed to list keys: invalid continue token: %v", err)
+		}
+		prefix = token.ContinueAt
+	} else {
+		prefix = ""
+	}
+	return names, prefix, nil
+}
+
+func (kc *KeyControl) refreshToken(ctx context.Context, renew time.Duration) {
```

Comment:
Can we add some comments to explain the renew logic for token? I understand it starts a timer and keeps halving the expiration time and the moment it reaches the time, next 5 mins it renews the token. Please explain in comment a bit.

Reply:
Done